contracts/reserve/GoodReserveCDai.sol

```diff
@@ -306,36 +455,67 @@ contract GoodReserveCDai is
 					getMarketMaker(),
 					this,
 					msg.sender,
-					_sellTo,
-					_gdAmount.sub(discount)
+					sellParameters.sellTo,
+					sellParameters.gdAmount.sub(discount)
 				);
 
 		uint256 tokenReturn =
 			getMarketMaker().sellWithContribution(
-				_sellTo,
-				_gdAmount,
+				sellParameters.sellTo,
+				sellParameters.gdAmount,
 				contributionAmount
-			);
-		require(
-			tokenReturn >= _minReturn,
-			"Token return must be above the minReturn"
 		);
-		require(
-			_sellTo.transfer(msg.sender, tokenReturn) == true,
-			"Transfer failed"
-		);
-
-		emit TokenSold(
-			msg.sender,
-			address(_sellTo),
-			_gdAmount,
-			contributionAmount,
-			_minReturn,
-			tokenReturn
-		);
-		return tokenReturn;
+		if(_returnInCDai == true){
+			address receiver = sellParameters.targetAddress == address(0x0) ? msg.sender : sellParameters.targetAddress ;
+			require(
+				tokenReturn >= _minReturn,
+				"Token return must be above the minReturn"
+			);
+			
+			require(
+				sellTo.transfer(receiver, tokenReturn) == true,
+				"Transfer failed"
+			);
+			
+			emit TokenSold(
+				receiver,
+				address(sellParameters.sellTo),
+				sellParameters.gdAmount,
+				contributionAmount,
+				sellParameters.minReturn,
+				tokenReturn
+			);
+			return (tokenReturn, contributionAmount);
+		} else {
+			cERC20 cDai = cERC20(nameService.getAddress("CDAI"));
+			ERC20 dai = ERC20(nameService.getAddress("DAI"));
+			
+			uint256 currDaiBalance = dai.balanceOf(address(this));
+			
+			uint256 daiResult = cDai.redeem(tokenReturn);
+			uint256 daiReturnAmount = (dai.balanceOf(address(this))).sub(currDaiBalance);
+			if (sellParameters.withTransfer == true){
+				address receiver = sellParameters.targetAddress == address(0x0) ? msg.sender : sellParameters.targetAddress ;
+				require(
+					dai.transfer(receiver, daiReturnAmount) == true,
+					"Transfer failed"
+				);
+				emit TokenSold(
+					receiver,
+					address(dai),
+					sellParameters.gdAmount,
+					contributionAmount,
+					sellParameters.minReturn,
+					daiReturnAmount
+				);
+				return (daiReturnAmount, contributionAmount);
```

Comment:
tokensold twice... DRY
this method should always return cdai and not do any transfer/redeem